src/tools.rs

```diff
@@ -654,6 +655,32 @@ pub(crate) fn single_value<T>(collection: impl IntoIterator<Item = T>) -> Option
     None
 }
 
+/// Compressor/decompressor buffer size.
+const BROTLI_BUFSZ: usize = 4096;
+
+/// Compresses `buf` to `Vec` using `brotli`.
+pub(crate) fn buf_compress(buf: &[u8]) -> Result<Vec<u8>> {
+    if buf.is_empty() {
+        return Ok(Vec::new());
+    }
+    let q: u32 = 11; // Use best compression level.
+    let lgwin: u32 = 22; // log2(LZ77 window size), it's the default for brotli CLI tool.
+    let mut compressor = brotli::CompressorWriter::new(Vec::new(), BROTLI_BUFSZ, q, lgwin);
+    compressor.write_all(buf)?;
+    Ok(compressor.into_inner())
+}
+
+/// Decompresses `buf` to `Vec` using `brotli`.
+pub(crate) fn buf_decompress(buf: &[u8]) -> Result<Vec<u8>> {
+    if buf.is_empty() {
```

Comment:
I'm not sure. We keep in the db messages deleted locally to be able to delete them later from the server also. So, i'd avoid calling `DecompressorWriter::new()` if not needed. And if we leave this check here, i'd leave it also in `buf_compress()` for symmetry

Reply:
`MsgId::trash()` doesn't compress the message, it just sets `mime_headers=''`, and then if we call `get_mime_headers()` for that message, it will call `buf_decompress()`. So, looks like it's really possible that `buf_decompress()` is called on an empty `buf`. And this [de]compression functions may be used for other things in the future, so i'd suggest to leave this check here

Reply:
Maybe it looks strange, yes, that we don't pass to brotli decompressor an empty buffer too, but i think it's ok to introduce such an empty placeholder value, i checked -- brotli returns an error if an empty file is passed to `brotli -d` and it's unlikely that they later will use an empty string as a special value for the most frequent word in the world f.e., it would be a sabotage.

Ok, i will change the code so that `buf_compress()` is never called for empty mime headers. Instead, for them `mime_compressed` column will be 0. (but i don't think this actually fixes anything)

Reply:
For now i left this as is, because:
- If we compress all mime headers incl. empty ones, the latter turn into 2-byte strings by brotli. Not nice.
- If compress only nonempty mime headers, we can't rely on `mime_compressed` column because it will be dropped eventually. So, we should rely on other things like whether a message is trashed or not. Looks complicated.

That's why i suggest to leave this empty string special value. Really it will never break, i don't think brotli devs are diversionists. But i can move this logic out of `buf_[de]compress()` so that it relates only to mime headers.

Reply:
Added a comment on this. I think that brotli compresses an empty buffer in the common way for the following reasons:
- There's just a common code with no exceptions.
- Empty buffers aren't a common case generally. This case is specific to us.
- To protect from truncation. But this is not valid for us, we don't protect from truncation of uncompressed data, so why protect for compressed one?